### PR TITLE
Fix ignore_failure behavior in _simulate?verbose and more cleanup

### PR DESCRIPTION
**Pull request by talevy**
- fix it so that processors with the `ignore_failure` option do not
  record their exception in the response
- add more tests to make empty `on_failure` block behavior more explicit

🎩  tip to @BigFunger for reporting these issues!

### CODE REVIEW

**Comment by talevy** [comment]
@martijnvg Looks like we missed this bug regarding `ignore_failure`.

This PR contains a fix for ignoring exceptions caused by `actualProcessor`s whose parent CompoundProcessor has `ignoreFailure = true`.

While diving into this issue, I also realized there is an inconsistency regarding the differences between an empty `on_failure` block and `ignore_failure: true`. Looks like they do not behave the same way.

should we reject configurations with empty `on_failure` blocks?
behavior could lead to confusion. I've added tests to reflect the current behavior, and I would like to know if this should be changed.

File: core/src/main/java/org/elasticsearch/common/io/stream/BytesStreamOutput.java [hunk 1]
```diff
@@ -21,6 +21,7 @@
 
 import org.elasticsearch.common.bytes.BytesReference;
 import org.elasticsearch.common.bytes.PagedBytesReference;
+import org.elasticsearch.common.inject.internal.Strings;
```
Comment (martijnvg) [inline core/src/main/java/org/elasticsearch/common/io/stream/BytesStreamOutput.java]: why is this file changed?

File: core/src/test/java/org/elasticsearch/action/ingest/SimulateExecutionServiceTests.java [hunk 2]
```diff
@@ -159,6 +159,27 @@ public void testExecuteVerboseItemWithOnFailure() throws Exception {
...
+        assertThat(testProcessor.getInvokedCounter(), equalTo(1));
+        assertThat(actualItemResponse, instanceOf(SimulateDocumentVerboseResult.class));
+        SimulateDocumentVerboseResult simulateDocumentVerboseResult = (SimulateDocumentVerboseResult) actualItemResponse;
+        assertThat(simulateDocumentVerboseResult.getProcessorResults().size(), equalTo(1));
+        assertThat(simulateDocumentVerboseResult.getProcessorResults().get(0).getProcessorTag(), equalTo("processor_0"));
+        assertThat(simulateDocumentVerboseResult.getProcessorResults().get(0).getFailure(), nullValue());
+        assertThat(simulateDocumentVerboseResult.getProcessorResults().get(0).getIngestDocument(), not(sameInstance(ingestDocument)));
+        assertIngestDocument(simulateDocumentVerboseResult.getProcessorResults().get(0).getIngestDocument(), ingestDocument);
+        assertThat(simulateDocumentVerboseResult.getProcessorResults().get(0).getIngestDocument().getSourceAndMetadata(), not(sameInstance(ingestDocument.getSourceAndMetadata())));
+//        assertThat(simulateDocumentVerboseResult.getProcessorResults().get(1).getProcessorTag(), equalTo("processor_1"));
```
Comment (martijnvg) [inline core/src/test/java/org/elasticsearch/action/ingest/SimulateExecutionServiceTests.java]: why is the commented out?

**Comment by martijnvg** [comment]
> Looks like we missed this bug regarding ignore_failure.

I completely forgot to check how this setting affects the simulate api, thanks for fixing! 

In my mind I thought I had this covered, but because of how the tracking processors get decorated
the tracking processor doesn't know about the ignore flag in the CompoundProcessor.

I think your fix looks good. Not sure how to do this differently with how currently the processor get decorated.

> should we reject configurations with empty on_failure blocks?

+1 I think this makes sense.

File: core/src/main/java/org/elasticsearch/common/io/stream/BytesStreamOutput.java (hunk 1 above)
Comment (talevy) [inline core/src/main/java/org/elasticsearch/common/io/stream/BytesStreamOutput.java]: woops, unrelated. I was exploring how serialization worked, this got left over

File: core/src/test/java/org/elasticsearch/action/ingest/SimulateExecutionServiceTests.java (hunk 2 above)
Comment (talevy) [inline core/src/test/java/org/elasticsearch/action/ingest/SimulateExecutionServiceTests.java]: unrelated to the test, will be removed

**Comment by talevy** [comment]
@martijnvg updated to reject empty `on_failure` from pipelines and processors.

File: core/src/main/java/org/elasticsearch/ingest/ConfigurationUtils.java [hunk 3]
```diff
@@ -256,6 +257,11 @@ private static Processor readProcessor(ProcessorsRegistry processorRegistry, Str
             List<Processor> onFailureProcessors = readProcessorConfigs(onFailureProcessorConfigs, processorRegistry);
             Processor processor;
             processor = factory.create(config);
+
+            if (onFailureProcessorConfigs != null && onFailureProcessors.isEmpty() == true) {
```
Comment (martijnvg) [inline core/src/main/java/org/elasticsearch/ingest/ConfigurationUtils.java]: instead of `onFailureProcessors.isEmpty() == true`, just `onFailureProcessors.isEmpty()`?

File: core/src/test/java/org/elasticsearch/ingest/PipelineFactoryTests.java [hunk 4]
```diff
@@ -86,6 +86,38 @@ public void testCreateWithPipelineOnFailure() throws Exception {
         assertThat(pipeline.getOnFailureProcessors().get(0).getType(), equalTo("test-processor"));
     }
 
+    public void testCreateWithPipelineEmptyOnFailure() throws Exception {
```
Comment (martijnvg) [inline core/src/test/java/org/elasticsearch/ingest/PipelineFactoryTests.java]: These tests will now, right? Because of the empty on_failure processors? I think we should expect an exception here instead.

Comment (talevy) [inline core/src/test/java/org/elasticsearch/ingest/PipelineFactoryTests.java]: ah, yes. forgot

File: core/src/main/java/org/elasticsearch/ingest/ConfigurationUtils.java (hunk 3 above)
Comment (talevy) [inline core/src/main/java/org/elasticsearch/ingest/ConfigurationUtils.java]: I prefer that as well, but for some reason remembered someone telling me to change it to `== true` since it is less error prone. I noticed it was used for config checking `config.isEmpty() == true`

**Comment by martijnvg** [comment]
LGTM
